### ISSUE TITLE
Getting false positive for "recover must be called inside a deferred function"

### ISSUE THREAD

**Issue by Integralist** [mgechev/revive#1348]
```go
// I have code that looks like this
defer consumer.PanicRecovery()

// Elsewhere my definition of PanicRecovery
func PanicRecovery() {
    if r := recover(); r != nil { // recover() is called here
        // ...
    }
}
```

This code works as I would expect it to, so I'm confused by revive sees it as problematic?

Is the expectation to do something like this...

```go
defer func() {
	if r := recover(); r != nil {
		PanicRecovery()
	}
}()
```

**Comment by chavacava** [comment]
Hi @Integralist thansk for creating the issue.

The case you describe, as you noticed, is not well handled by `revive`. The rule that checks uses of `recover` has not "global information" on how and where functions are used; it does its best by analyzing calls to `recover` in the very local context of the calling function. In the case you describe, the rule detects that `PanicRecovery` calls `recover` but is not aware that `PanicRecovery` is deferred elsewhere. As almost always in this kind of tools, it's a trade-of between correctness and complexity: `revive` opts by being _almost always correct_ and simple.

As mentioned by @denisvmedia, the case is not very "go idiomatic" (also notice there is no guaranty of `PanicRecovery` being **always** called in the context of a defer)

I'll will close the issue. If you (or others) think that the problem must be addressed (i.e. fixed somehow) please do not hesitate to re-open the discussion.
